File: c/command.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* ============================================================
 * Command Pattern in Pure C — Trade Management with Undo/Redo
 *
 * Two approaches shown:
 *
 * 1. Tagged Union (closed set of commands)
 *    - Commands are pure data — no behavior embedded
 *    - Execute/undo are free functions that switch on the tag
 *    - History is a plain array — trivially copyable/snapshottable
 *
 * 2. Function Pointer Table (open extension)
 *    - Each command carries its own execute/undo function pointers
 *    - New command types can be added without modifying existing code
 *    - Still value-semantic with inline storage
 *
 * In both cases, the command history is fully copyable.
 * ============================================================ */

/* --- Receiver: Portfolio --- */

#define MAX_POSITIONS 32

typedef struct {
    char symbols[MAX_POSITIONS][16];
    int quantities[MAX_POSITIONS];
    int count;
    double cash;
} Portfolio;

static void portfolio_init(Portfolio *p, double cash) {
    memset(p, 0, sizeof(*p));
    p->cash = cash;
}

static int portfolio_find(Portfolio *p, const char *symbol) {
    for (int i = 0; i < p->count; ++i) {
        if (strcmp(p->symbols[i], symbol) == 0) return i;
    }
    return -1;
}

static void portfolio_adjust(Portfolio *p, const char *symbol, int qty_delta,
                              double cash_delta) {
    int idx = portfolio_find(p, symbol);
    if (idx < 0) {
        idx = p->count++;
        strncpy(p->symbols[idx], symbol, 15);
        p->symbols[idx][15] = '\0';
        p->quantities[idx] = 0;
    }
    p->quantities[idx] += qty_delta;
    p->cash += cash_delta;
}

static void portfolio_buy(Portfolio *p, const char *symbol, int qty, double price) {
    portfolio_adjust(p, symbol, qty, -(qty * price));
    printf("  [EXEC] BUY  %d %s @ $%.2f  (cash: $%.2f)\n",
           qty, symbol, price, p->cash);
}

static void portfolio_sell(Portfolio *p, const char *symbol, int qty, double price) {
    portfolio_adjust(p, symbol, -qty, qty * price);
    printf("  [EXEC] SELL %d %s @ $%.2f  (cash: $%.2f)\n",
           qty, symbol, price, p->cash);
}

static void portfolio_reverse_buy(Portfolio *p, const char *symbol,
                                    int qty, double price) {
    portfolio_adjust(p, symbol, -qty, qty * price);
    printf("  [UNDO] BUY  %d %s @ $%.2f reversed  (cash: $%.2f)\n",
           qty, symbol, price, p->cash);
}

static void portfolio_reverse_sell(Portfolio *p, const char *symbol,
                                     int qty, double price) {
    portfolio_adjust(p, symbol, qty, -(qty * price));
    printf("  [UNDO] SELL %d %s @ $%.2f reversed  (cash: $%.2f)\n",
           qty, symbol, price, p->cash);
}

static void portfolio_print(const Portfolio *p) {
    puts("  Portfolio:");
    printf("    Cash: $%.2f\n", p->cash);
    for (int i = 0; i < p->count; ++i) {
        if (p->quantities[i] != 0)
            printf("    %s: %d shares\n", p->symbols[i], p->quantities[i]);
    }
}

/* ============================================================
 * APPROACH 1: Tagged Union Commands
 * ============================================================ */

typedef enum {
    CMD_BUY,
    CMD_SELL
} TradeCommandType;

typedef struct {
    TradeCommandType type;
    char symbol[16];
    int quantity;
    double price;
} TradeCommand;

static TradeCommand cmd_buy(const char *symbol, int qty, double price) {
    TradeCommand cmd;
    cmd.type = CMD_BUY;
    strncpy(cmd.symbol, symbol, 15);
    cmd.symbol[15] = '\0';
    cmd.quantity = qty;
    cmd.price = price;
    return cmd;
}

static TradeCommand cmd_sell(const char *symbol, int qty, double price) {
    TradeCommand cmd;
    cmd.type = CMD_SELL;
    strncpy(cmd.symbol, symbol, 15);
    cmd.symbol[15] = '\0';
    cmd.quantity = qty;
    cmd.price = price;
    return cmd;
}

static void cmd_execute(const TradeCommand *cmd, Portfolio *p) {
    switch (cmd->type) {
    case CMD_BUY:  portfolio_buy(p, cmd->symbol, cmd->quantity, cmd->price);  break;
    case CMD_SELL: portfolio_sell(p, cmd->symbol, cmd->quantity, cmd->price); break;
    }
}

static void cmd_undo(const TradeCommand *cmd, Portfolio *p) {
    switch (cmd->type) {
    case CMD_BUY:  portfolio_reverse_buy(p, cmd->symbol, cmd->quantity, cmd->price);  break;
    case CMD_SELL: portfolio_reverse_sell(p, cmd->symbol, cmd->quantity, cmd->price); break;
    }
}

static void cmd_describe(const TradeCommand *cmd, char *buf, size_t len) {
    snprintf(buf, len, "%s %d %s @ $%.2f",
             cmd->type == CMD_BUY ? "BUY" : "SELL",
             cmd->quantity, cmd->symbol, cmd->price);
}

/* --- Command History: plain array of values --- */

#define MAX_HISTORY 64

typedef struct {
    TradeCommand executed[MAX_HISTORY];
    int exec_count;
    TradeCommand undone[MAX_HISTORY];
    int undo_count;
} TradeHistory;

static void history_init(TradeHistory *h) {
    h->exec_count = 0;
    h->undo_count = 0;
}

static void history_execute(TradeHistory *h, TradeCommand cmd, Portfolio *p) {
    cmd_execute(&cmd, p);
    h->executed[h->exec_count++] = cmd;  /* value copy */
    h->undo_count = 0;  /* clear redo stack */
}

static int history_undo(TradeHistory *h, Portfolio *p) {
    if (h->exec_count == 0) return 0;
    TradeCommand cmd = h->executed[--h->exec_count];
    cmd_undo(&cmd, p);
    h->undone[h->undo_count++] = cmd;
    return 1;
}

static int history_redo(TradeHistory *h, Portfolio *p) {
    if (h->undo_count == 0) return 0;
    TradeCommand cmd = h->undone[--h->undo_count];
    cmd_execute(&cmd, p);
    h->executed[h->exec_count++] = cmd;
    return 1;
}

static void history_print(const TradeHistory *h) {
    char buf[128];
    puts("  Trade History:");
    for (int i = 0; i < h->exec_count; ++i) {
        cmd_describe(&h->executed[i], buf, sizeof(buf));
        printf("    %d. %s\n", i + 1, buf);
    }
    if (h->exec_count == 0) puts("    (empty)");
}

/* ============================================================
 * APPROACH 2: Function Pointer Table (Type Erasure)
 * ============================================================ */

#define ERASED_CMD_STORAGE 48

typedef struct {
    void (*execute)(const void *self, Portfolio *p);
    void (*undo)(const void *self, Portfolio *p);
    void (*describe)(const void *self, char *buf, size_t len);
    char storage[ERASED_CMD_STORAGE];
} ErasedCommand;

/* --- Market Buy command --- */

typedef struct {
    char symbol[16];
    int quantity;
    double price;
} MarketBuyData;

static void market_buy_exec(const void *self, Portfolio *p) {
    const MarketBuyData *d = (const MarketBuyData *)self;
    portfolio_buy(p, d->symbol, d->quantity, d->price);
}

static void market_buy_undo(const void *self, Portfolio *p) {
    const MarketBuyData *d = (const MarketBuyData *)self;
    portfolio_reverse_buy(p, d->symbol, d->quantity, d->price);
}

static void market_buy_describe(const void *self, char *buf, size_t len) {
    const MarketBuyData *d = (const MarketBuyData *)self;
    snprintf(buf, len, "MARKET BUY %d %s @ $%.2f", d->quantity, d->symbol, d->price);
}

static ErasedCommand erased_market_buy(const char *symbol, int qty, double price) {
    ErasedCommand cmd;
    cmd.execute = market_buy_exec;
    cmd.undo = market_buy_undo;
    cmd.describe = market_buy_describe;
    MarketBuyData *d = (MarketBuyData *)cmd.storage;
    strncpy(d->symbol, symbol, 15);
    d->symbol[15] = '\0';
    d->quantity = qty;
    d->price = price;
    return cmd;
}

/* --- Limit Sell command --- */

typedef struct {
    char symbol[16];
    int quantity;
    double limit_price;
} LimitSellData;

static void limit_sell_exec(const void *self, Portfolio *p) {
    const LimitSellData *d = (const LimitSellData *)self;
    portfolio_sell(p, d->symbol, d->quantity, d->limit_price);
}

static void limit_sell_undo(const void *self, Portfolio *p) {
    const LimitSellData *d = (const LimitSellData *)self;
    portfolio_reverse_sell(p, d->symbol, d->quantity, d->limit_price);
}

static void limit_sell_describe(const void *self, char *buf, size_t len) {
    const LimitSellData *d = (const LimitSellData *)self;
    snprintf(buf, len, "LIMIT SELL %d %s @ $%.2f",
             d->quantity, d->symbol, d->limit_price);
}

static ErasedCommand erased_limit_sell(const char *symbol, int qty, double price) {
    ErasedCommand cmd;
    cmd.execute = limit_sell_exec;
    cmd.undo = limit_sell_undo;
    cmd.describe = limit_sell_describe;
    LimitSellData *d = (LimitSellData *)cmd.storage;
    strncpy(d->symbol, symbol, 15);
    d->symbol[15] = '\0';
    d->quantity = qty;
    d->limit_price = price;
    return cmd;
}

/* ============================================================ */

int main(void) {
    puts("=== C Command Pattern: Trade Management ===");
    puts("========== Approach 1: Tagged Union ==========\n");

    Portfolio portfolio;
    portfolio_init(&portfolio, 1000000.0);

    TradeHistory history;
    history_init(&history);

    puts("--- Executing trades ---");
    history_execute(&history, cmd_buy("AAPL", 100, 185.50), &portfolio);
    history_execute(&history, cmd_buy("GOOGL", 50, 140.25), &portfolio);
    history_execute(&history, cmd_sell("MSFT", 75, 420.00), &portfolio);

    puts("");
    portfolio_print(&portfolio);
    puts("");
    history_print(&history);

    puts("\n--- Undo last trade ---");
    history_undo(&history, &portfolio);
    portfolio_print(&portfolio);

    puts("\n--- Undo another ---");
    history_undo(&history, &portfolio);
    portfolio_print(&portfolio);

    puts("\n--- Redo ---");
    history_redo(&history, &portfolio);
    portfolio_print(&portfolio);

    /* Snapshot: just copy the structs */
    puts("\n--- Snapshot history (plain struct copy!) ---");
    TradeHistory snapshot = history;  /* entire history copied by value */
    printf("  Snapshot has %d trades\n", snapshot.exec_count);

    /* Continue on original */
    history_execute(&history, cmd_sell("AAPL", 50, 190.00), &portfolio);

    puts("\n--- Original history ---");
    history_print(&history);

    puts("\n--- Snapshot unchanged ---");
    history_print(&snapshot);

    printf("\n  sizeof(TradeCommand) = %zu bytes\n", sizeof(TradeCommand));
    printf("  sizeof(TradeHistory) = %zu bytes (all on stack)\n", sizeof(TradeHistory));

    /* ============================================================ */
    puts("\n========== Approach 2: Function Pointer Table ==========\n");

    Portfolio portfolio2;
    portfolio_init(&portfolio2, 500000.0);

    ErasedCommand commands[2];
    commands[0] = erased_market_buy("TSLA", 200, 175.00);
    commands[1] = erased_limit_sell("NVDA", 30, 890.50);

    puts("--- Executing type-erased commands ---");
    for (int i = 0; i < 2; ++i) {
        commands[i].execute(commands[i].storage, &portfolio2);
    }

    puts("\n--- Undoing all ---");
    for (int i = 1; i >= 0; --i) {
        commands[i].undo(commands[i].storage, &portfolio2);
    }

    portfolio_print(&portfolio2);

    /* Copy erased commands */
    puts("\n--- Erased commands are copyable ---");
    ErasedCommand cmd_copy = commands[0];
    char buf[128];
    cmd_copy.describe(cmd_copy.storage, buf, sizeof(buf));
    printf("  Copied command: %s\n", buf);

    printf("\n  sizeof(ErasedCommand) = %zu bytes\n", sizeof(ErasedCommand));

    return 0;
}

File: rust/src/bin/command.rs
// ============================================================
// Command Pattern in Rust — Trade Management with Undo/Redo
//
// Two approaches:
//
// 1. Enum commands (closed set)
//    - Commands are pure data, #[derive(Clone)]
//    - Execute/undo are match arms in free functions
//    - History is Vec<TradeAction> — trivially cloneable
//    - Perfect for event sourcing and audit trails
//
// 2. Trait objects (open extension)
//    - Box<dyn Command> with clone_box for value semantics
//    - New command types don't require modifying existing code
//    - Slightly more ceremony than enum approach
//
// Rust's ownership system gives a unique advantage here:
// the borrow checker ensures commands can't hold dangling
// references to the portfolio. Commands store data, not
// references — the portfolio is passed explicitly to execute/undo.
// ============================================================

use std::collections::HashMap;
use std::fmt;

// --- Receiver: Portfolio ---

#[derive(Debug, Clone)]
struct Portfolio {
    positions: HashMap<String, i32>,
    cash: f64,
}

impl Portfolio {
    fn new(cash: f64) -> Self {
        Self {
            positions: HashMap::new(),
            cash,
        }
    }

    fn buy(&mut self, symbol: &str, qty: i32, price: f64) {
        *self.positions.entry(symbol.to_string()).or_insert(0) += qty;
        self.cash -= qty as f64 * price;
        println!(
            "  [EXEC] BUY  {} {} @ ${:.2}  (cash: ${:.2})",
            qty, symbol, price, self.cash
        );
    }

    fn sell(&mut self, symbol: &str, qty: i32, price: f64) {
        *self.positions.entry(symbol.to_string()).or_insert(0) -= qty;
        self.cash += qty as f64 * price;
        println!(
            "  [EXEC] SELL {} {} @ ${:.2}  (cash: ${:.2})",
            qty, symbol, price, self.cash
        );
    }

    fn reverse_buy(&mut self, symbol: &str, qty: i32, price: f64) {
        *self.positions.entry(symbol.to_string()).or_insert(0) -= qty;
        self.cash += qty as f64 * price;
        println!(
            "  [UNDO] BUY  {} {} @ ${:.2} reversed  (cash: ${:.2})",
            qty, symbol, price, self.cash
        );
    }

    fn reverse_sell(&mut self, symbol: &str, qty: i32, price: f64) {
        *self.positions.entry(symbol.to_string()).or_insert(0) += qty;
        self.cash -= qty as f64 * price;
        println!(
            "  [UNDO] SELL {} {} @ ${:.2} reversed  (cash: ${:.2})",
            qty, symbol, price, self.cash
        );
    }

    fn print_positions(&self) {
        println!("  Portfolio:");
        println!("    Cash: ${:.2}", self.cash);
        for (sym, qty) in &self.positions {
            if *qty != 0 {
                println!("    {}: {} shares", sym, qty);
            }
        }
    }
}

// ============================================================
// APPROACH 1: Enum Commands (closed set)
// ============================================================

#[derive(Debug, Clone)]
enum TradeAction {
    Buy {
        symbol: String,
        quantity: i32,
        price: f64,
    },
    Sell {
        symbol: String,
        quantity: i32,
        price: f64,
    },
}

impl fmt::Display for TradeAction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Buy {
                symbol,
                quantity,
                price,
            } => write!(f, "BUY {} {} @ ${:.2}", quantity, symbol, price),
            Self::Sell {
                symbol,
                quantity,
                price,
            } => write!(f, "SELL {} {} @ ${:.2}", quantity, symbol, price),
        }
    }
}

impl TradeAction {
    fn execute(&self, portfolio: &mut Portfolio) {
        match self {
            Self::Buy {
                symbol,
                quantity,
                price,
            } => portfolio.buy(symbol, *quantity, *price),
            Self::Sell {
                symbol,
                quantity,
                price,
            } => portfolio.sell(symbol, *quantity, *price),
        }
    }

    fn undo(&self, portfolio: &mut Portfolio) {
        match self {
            Self::Buy {
                symbol,
                quantity,
                price,
            } => portfolio.reverse_buy(symbol, *quantity, *price),
            Self::Sell {
                symbol,
                quantity,
                price,
            } => portfolio.reverse_sell(symbol, *quantity, *price),
        }
    }
}

// --- Command History: Vec of Clone-able values ---

#[derive(Debug, Clone)]
struct TradeHistory {
    executed: Vec<TradeAction>,
    undone: Vec<TradeAction>,
}

impl TradeHistory {
    fn new() -> Self {
        Self {
            executed: Vec::new(),
            undone: Vec::new(),
        }
    }

    fn execute(&mut self, action: TradeAction, portfolio: &mut Portfolio) {
        action.execute(portfolio);
        self.executed.push(action);
        self.undone.clear();
    }

    fn undo(&mut self, portfolio: &mut Portfolio) -> bool {
        if let Some(action) = self.executed.pop() {
            action.undo(portfolio);
            self.undone.push(action);
            true
        } else {
            false
        }
    }

    fn redo(&mut self, portfolio: &mut Portfolio) -> bool {
        if let Some(action) = self.undone.pop() {
            action.execute(portfolio);
            self.executed.push(action);
            true
        } else {
            false
        }
    }

    fn print_history(&self) {
        println!("  Trade History:");
        if self.executed.is_empty() {
            println!("    (empty)");
        } else {
            for (i, action) in self.executed.iter().enumerate() {
                println!("    {}. {}", i + 1, action);
            }
        }
    }
}

// ============================================================
// APPROACH 2: Trait Objects (open extension)
// ============================================================

trait Command: fmt::Debug {
    fn execute(&self, portfolio: &mut Portfolio);
    fn undo(&self, portfolio: &mut Portfolio);
    fn description(&self) -> String;
    fn clone_box(&self) -> Box<dyn Command>;
}

impl Clone for Box<dyn Command> {
    fn clone(&self) -> Self {
        self.clone_box()
    }
}

#[derive(Debug, Clone)]
struct MarketBuy {
    symbol: String,
    quantity: i32,
    price: f64,
}

impl Command for MarketBuy {
    fn execute(&self, portfolio: &mut Portfolio) {
        portfolio.buy(&self.symbol, self.quantity, self.price);
    }

    fn undo(&self, portfolio: &mut Portfolio) {
        portfolio.reverse_buy(&self.symbol, self.quantity, self.price);
    }

    fn description(&self) -> String {
        format!(
            "MARKET BUY {} {} @ ${:.2}",
            self.quantity, self.symbol, self.price
        )
    }

    fn clone_box(&self) -> Box<dyn Command> {
        Box::new(self.clone())
    }
}

#[derive(Debug, Clone)]
struct LimitSell {
    symbol: String,
    quantity: i32,
    limit_price: f64,
}

impl Command for LimitSell {
    fn execute(&self, portfolio: &mut Portfolio) {
        portfolio.sell(&self.symbol, self.quantity, self.limit_price);
    }

    fn undo(&self, portfolio: &mut Portfolio) {
        portfolio.reverse_sell(&self.symbol, self.quantity, self.limit_price);
    }

    fn description(&self) -> String {
        format!(
            "LIMIT SELL {} {} @ ${:.2}",
            self.quantity, self.symbol, self.limit_price
        )
    }

    fn clone_box(&self) -> Box<dyn Command> {
        Box::new(self.clone())
    }
}

// ============================================================

fn main() {
    println!("=== Rust Command Pattern: Trade Management ===");
    println!("========== Approach 1: Enum Commands ==========\n");

    let mut portfolio = Portfolio::new(1_000_000.0);
    let mut history = TradeHistory::new();

    println!("--- Executing trades ---");
    history.execute(
        TradeAction::Buy {
            symbol: "AAPL".into(),
            quantity: 100,
            price: 185.50,
        },
        &mut portfolio,
    );
    history.execute(
        TradeAction::Buy {
            symbol: "GOOGL".into(),
            quantity: 50,
            price: 140.25,
        },
        &mut portfolio,
    );
    history.execute(
        TradeAction::Sell {
            symbol: "MSFT".into(),
            quantity: 75,
            price: 420.00,
        },
        &mut portfolio,
    );

    println!();
    portfolio.print_positions();
    println!();
    history.print_history();

    println!("\n--- Undo last trade ---");
    history.undo(&mut portfolio);
    portfolio.print_positions();

    println!("\n--- Undo another ---");
    history.undo(&mut portfolio);
    portfolio.print_positions();

    println!("\n--- Redo ---");
    history.redo(&mut portfolio);
    portfolio.print_positions();

    // Snapshot: just clone
    println!("\n--- Snapshot history (clone!) ---");
    let snapshot = history.clone();
    println!("  Snapshot has {} trades", snapshot.executed.len());

    // Continue on original
    history.execute(
        TradeAction::Sell {
            symbol: "AAPL".into(),
            quantity: 50,
            price: 190.00,
        },
        &mut portfolio,
    );

    println!("\n--- Original history ---");
    history.print_history();

    println!("\n--- Snapshot unchanged ---");
    snapshot.print_history();

    // ============================================================
    println!("\n========== Approach 2: Trait Objects ==========\n");

    let mut portfolio2 = Portfolio::new(500_000.0);

    let commands: Vec<Box<dyn Command>> = vec![
        Box::new(MarketBuy {
            symbol: "TSLA".into(),
            quantity: 200,
            price: 175.00,
        }),
        Box::new(LimitSell {
            symbol: "NVDA".into(),
            quantity: 30,
            limit_price: 890.50,
        }),
    ];

    println!("--- Executing trait commands ---");
    for cmd in &commands {
        cmd.execute(&mut portfolio2);
    }

    println!("\n--- Undoing all ---");
    for cmd in commands.iter().rev() {
        cmd.undo(&mut portfolio2);
    }

    portfolio2.print_positions();

    // Commands are cloneable
    println!("\n--- Commands are cloneable ---");
    let commands_copy = commands.clone();
    println!("  Original: {} commands", commands.len());
    println!("  Copy:     {} commands", commands_copy.len());
    for cmd in &commands_copy {
        println!("    {}", cmd.description());
    }

    // ============================================================
    // Rust's ownership advantage:
    //
    // Notice that commands DON'T hold references to the portfolio.
    // The portfolio is passed as &mut to execute/undo. The borrow
    // checker enforces this at compile time — you literally cannot
    // create a command that holds a dangling reference to a portfolio
    // that might be moved or dropped. This eliminates an entire
    // class of bugs that C++ and C must manage manually.
    // ============================================================
}
